File: 253.meeting-rooms-ii.py
#
# @lc app=leetcode id=253 lang=python3
#
# [253] Meeting Rooms II
#
# https://leetcode.com/problems/meeting-rooms-ii/description/
#
# algorithms
# Medium (41.81%)
# Total Accepted:    121.4K
# Total Submissions: 290.3K
# Testcase Example:  '[[0,30],[5,10],[15,20]]'
#
# Given an array of meeting time intervals consisting of start and end times
# [[s1,e1],[s2,e2],...] (si < ei), find the minimum number of conference rooms
# required.
#
# Example 1:
#
#
# Input: [[0, 30],[5, 10],[15, 20]]
# Output: 2
#
# Example 2:
#
#
# Input: [[7,10],[2,4]]
# Output: 1
#
#
# Definition for an interval.


class Solution(object):
    def minMeetingRooms(self, intervals):
        """
        :type intervals: List[Interval]
        :rtype: int
        """
        if (not intervals):
            return 0

        starting = {}
        ending = {}
        lst = []
        room = 0
        need = 0
        
        for item in intervals:
            if item.start in starting:
                starting[item.start] += 1
            else:
                starting[item.start] = 1
                if item.start not in ending:
                    lst.append(item.start)
            if item.end in ending:
                ending[item.end] += 1
            else:
                ending[item.end] = 1
                if item.end not in starting:
                    lst.append(item.end)
                    
        lst.sort()
                
        for item in lst:
            if item in starting:
                room += starting[item]
            if item in ending:
                room -= ending[item]
            if room > need:
                need = room
            
        return need

File: 253.meeting-rooms-ii.java
/*
 * @lc app=leetcode id=253 lang=java
 *
 * [253] Meeting Rooms II
 *
 * https://leetcode.com/problems/meeting-rooms-ii/description/
 *
 * algorithms
 * Medium (41.81%)
 * Total Accepted:    121.4K
 * Total Submissions: 290.3K
 * Testcase Example:  '[[0,30],[5,10],[15,20]]'
 *
 * Given an array of meeting time intervals consisting of start and end times
 * [[s1,e1],[s2,e2],...] (si < ei), find the minimum number of conference rooms
 * required.
 * 
 * Example 1:
 * 
 * 
 * Input: [[0, 30],[5, 10],[15, 20]]
 * 0 5 15 17
 * 10 20 22 30 
 * Output: 2
 * 
 * Example 2:
 * 
 * 
 * Input: [[7,10],[2,4]]
 * Output: 1
 * 
 */
/**
 * Definition for an interval.
 * public class Interval {
 *     int start;
 *     int end;
 *     Interval() { start = 0; end = 0; }
 *     Interval(int s, int e) { start = s; end = e; }
 * }
 */

import java.util.*;

class Solution {
    public int minMeetingRooms(Interval[] intervals) {
        int[] start = new int[intervals.length];
        int[] end = new int[intervals.length];
        for (int i = 0; i < start.length; i++) {
            start[i] = intervals[i].start;
            end[i] = intervals[i].end;
        }
        Arrays.sort(start);
        Arrays.sort(end);
        int rooms = 0, endsItr = 0;
        for (int i = 0; i < start.length; i++) {
            if (start[i] < end[endsItr]) {
                rooms++;
            } else {
                endsItr++;
            }
        }

        return rooms;

    }
}
